Project 9

sun4205 opened the pull request:
Hello, my name is Hanna.
It’s a pleasure to contact you.
This project was really tough, but I am glad I can finally submit it.
I hope there aren’t too many mistakes I need to fix.
Thank you for reviewing my project.
I am looking forward to seeing my mistakes so that I can learn. :)
Thank you.

sun4205 committed:
api first step

sun4205 committed:
editing avatar

sun4205 committed:
editing avatar

sun4205 committed:
delete modal

sun4205 committed:
delete,like cards

sun4205 committed:
like cards

sun4205 committed:
post

sun4205 committed:
updating delete form

sun4205 committed:
avatar

sun4205 committed:
avatar

sun4205 committed:
avatar

sun4205 committed:
final fix

sun4205 committed:
profile fix

sun4205 committed:
css change

sun4205 committed:
pakage change

gennady-bars commented on src/utils/Api.js:
```diff
@@ -0,0 +1,106 @@
+    return fetch(`${this._baseUrl}/cards`, {
+      headers: this._headers,
+    }).then((res) => {
+      if (res.ok) {
+        return res.json();
+      }
+      Promise.reject(`Error: ${res.status}`);
```
```js
.then(res => {
                if (res.ok) {
                    return res.json();
                }
                return Promise.reject(`Error ${res.status}`);
            })

```

The code of `ok` checking is repeated in **every** request of the project. To avoid it you need to make a special method `_checkResponse` that will check any response from the server if it’s ok or not.


```js
_checkResponse(res) {
        // here is the code of the checking
    }

```

And now you can replace redundant repeating with the only line:


```js
 .then(this._checkResponse)

```

Please, pay attention that you need to pass only the **reference** for the method rather than the call. 

**COULD BE IMPROVED**

You can make a special method (function ) for fetching and checking responses **not to duplicate** it in every request:

```js

  _request(url, options) {
    return fetch(url, options).then(this._checkResponse)
  }

```

And now you can safely replace `fetch` with `this._request` and remove repeated checking. And you don’t need to change anything else.

gennady-bars commented on src/pages/index.js:
```diff
@@ -1,5 +1,12 @@
+import Api from "../utils/Api.js";
+import { setButtonText } from "../utils/helpers.js";
 
 const initialCards = [
```
From the 8th sprint all constants (including `initialCards` and `settings`) should be in `constants.js` in the `utils` folder to clean up the main file.

gennady-bars commented on src/pages/index.js:
```diff
@@ -34,13 +41,38 @@ const initialCards = [
+
+api
+  .getAppInfo()
+  .then(([cards, userInfo]) => {
```
Good job that you get the whole info with `Promise.all`

gennady-bars commented on src/pages/index.js:
```diff
@@ -56,6 +88,19 @@ const cardModalClosebtn = cardModal.querySelector(".modal__close-btn");
+const avatarModal = document.getElementById("modal-avatar-edit");
+const avatarElement = document.querySelector(".profile__avatar");
+const avatarEditButton = document.querySelector(".profile__avatar-btn");
+const avatarForm = avatarModal.querySelector(".modal__form");
```
**COULD BE IMPROVED**

We can get any form using their **unique** attribute `name` or `id` from `document.forms`

```js

const profileForm = document.forms["profile-form"];
const cardForm = document.forms["card-form"];

```

You need to add `name` or `id` first in `index.html` to achieve that.

More info here https://developer.mozilla.org/en-US/docs/Web/API/Document/forms

gennady-bars commented on src/pages/index.js:
```diff
@@ -127,28 +193,101 @@ function getCardElement(data) {
-  profileDescription.textContent = editModalDescriptionInput.value;
-  closeModal(editModal);
+
+  const submitBtn = e.submitter;
+  setButtonText(submitBtn, true);
```
**COULD BE IMPROVED**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **resetting** and **catching errors**


```js
// define a function for changing the button text. It accepts 4 params (the 2 last are optional with default texts)
export function renderLoading(isLoading, button, buttonText='Save', loadingText='Saving...') {
  if (isLoading) {
    button.textContent = loadingText
  } else {
    button.textContent = buttonText
  }
}

// define a universal function that accepts a request function, event and a default loading text 
function handleSubmit(request, evt, loadingText = 'Saving...') {
 // You need to prevent the default action in any submit handler
  evt.preventDefault();

  // the button is always available inside `event` as `submitter`
  const submitButton = evt.submitter;
  // fix the initial button text
  const initialText = submitButton.textContent;
  // change the button text before requesting
  renderLoading(true, submitButton, initialText, loadingText);
  // call the request function to be able to use the promise chain
  request()
    .then(() => {
      // any form should be reset after a successful response 
      // evt.target is the form in any submit handler		
      evt.target.reset();
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
    .catch(console.error)      

    // and in finally we need to stop loading
    .finally(() => {
      renderLoading(false, submitButton, initialText);
    });
}

```

Here is an example of handling the profile form submit:

```js
function handleProfileFormSubmit(evt) {
  // create a request function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally`
    return editProfile(nameInput.value, jobInput.value).then((userData) => {
      userName.textContent = userData.name;
      userJob.textContent = userData.about;
    });
  }
  // here we call handleSubmit passing the request and event (if you want a different loading text then you need to pass the 3rd param)
  handleSubmit(makeRequest, evt);
}

```

So, this way you can remove a lot of code duplicating. You will not need to search buttons, pass initial button texts and so on.

`handleSubmit` and `renderLoading` should be placed in `utils.js`, because they are utility functions

sun4205 replied:
Thank you for the detail explanation. The code looks more concise and organized now. I really appreciate you letting me know. :)

gennady-bars commented on src/pages/index.js:
```diff
@@ -168,22 +307,15 @@ cardModalbtn.addEventListener("click", () => {
   openModal(cardModal);
```
```js

cardModalbtn.addEventListener("click", () => {
  editModalNameInput.value = profileName.textContent;
  editModalDescriptionInput.value = profileDescription.textContent;
  openModal(cardModal);
});

```


Please, delete the 2 lines of code (305 and 306) because they are for the **profile** opening

gennady-bars reviewed (commented):
**↓Expand to see the whole summary↓**

Hello, Hanna!

Great job on submitting your project 9! This time you’ve implemented `Api` integration. So you can create your own simple sites from now on! 

**Parts I really liked in your project:**

* Good job that you **close** modals only in blocks `then`
* Good job that you `catch` possible errors at the end of server requests.
* Good job that you return the button text in `finally`
* Good job that you get the whole info with `Promise.all`

There are few things that need to be corrected in your project. They're mostly minor issues that are easy to fix.

**Needs correcting:**

* All comments in the code  and the comments below:
* The code of `ok` checking is repeated in **every** request of the project. To avoid it you need to make a special method `_checkResponse` that will check any response from the server if it’s ok or not.
* From the 8th sprint all constants (including `initialCards` and `settings`) should be in `constants.js` in the `utils` folder to clean up the main file.
* Please, delete the 2 lines of code (305 and 306) because they are for the **profile** opening

There are some comments in your project that can help you make it even better. Please check the `Could be improved` comments.

**A reminder:** It’s necessary to fix all comments to get the project accepted. And `Could be improved` comments are optional, they will be up to you if you want to implement them.

Your hard work on this project is truly commendable. While there are still some adjustments needed. If you need any assistance with the feedback or understanding certain concepts, our tutors are here to help. Keep moving forward and maintain your dedication — you are doing an amazing job!✨

sun4205 committed:
project correction

sun4205 committed:
project correction

sun4205 committed:
project correction

sun4205 committed:
project correction

sun4205 commented:
Thank you for your detailed review. It had useful information, especially about the duplicated code. I fixed the code based on what you told me. Everything works fine, but there's one thing I don't understand. After submitting all the buttons — edit avatar, profile, new post, and delete — the screen doesn't update right away. After refreshing the page, the screen updates. I can't figure out why this is happening after I fixed my code. Please let me know what did I do wrong. Thank you again.

gennady-bars commented on src/utils/Api.js:
```diff
@@ -0,0 +1,81 @@
+    return Promise.reject(`Error: ${res.status}`);
+  }
+
+  _request(url, options) {
+    return fetch(url, options).then(this._checkResponse);
```
**COULD BE IMPROVED**

You can make a universal method `_request` not to duplicate the `ok` checking, `headers` and `baseUrl`


```js
 _request(endpoint, options = {}) {
    const finalOptions = {
      // add the default headers
      headers: this._headers,
      // spread the other options if provided
      ...options,
    };
      // add the endpoint to the base url
    const url = `${this._baseUrl}${endpoint}`;
    return fetch(url, finalOptions).then(this._checkResponse);
  }

  // usage

  getUserInfo() {
    // pass only the endpoint here
    return this._request("/users/me");
  }

  getInitialCards() {
    // pass only the endpoint here
    return this._request("/cards");
  }

  // and so on



```

gennady-bars commented on src/pages/index.js:
```diff
@@ -163,27 +180,52 @@ profileEditButton.addEventListener("click", () => {
+        name: editModalNameInput.value,
+        about: editModalDescriptionInput.value,
+      }),
+    e
```
` e`


You've forgotten to add the updating in `then` inside the request funcitions

```js

function handleEditForSubmit(e) {
  handleSubmit(
    () =>
      api
        .editUserInfo({
          name: editModalNameInput.value,
          about: editModalDescriptionInput.value,
        })
        .then(() => {
          // here you update the page
          closeModal(editModal);
        }),
    e
  );
}

```

gennady-bars reviewed (commented):
(**Please, expand the general comment ↓**)

Dear Hanna, 

unfortunately I have to reject the project without review:


* You've forgotten to add the updating in `then` inside the request functions
* I can't add new cards
* I can't delete cards
* I can't edit the profile data or change the avatar

gennady-bars commented:
* You've forgotten to add the updating in `then` inside the request funcitions

sun4205 committed:
project correction

sun4205 committed:
project correction

sun4205 committed:
project correction

sun4205 committed:
project correction

sun4205 commented:
Thank you for the comment.
This project is taking a long time to finish, but that's okay. 
It's part of the learning process.
This is my revised code. 
Thank you. :)

gennady-bars commented on src/pages/index.js:
```diff
@@ -163,27 +181,82 @@ profileEditButton.addEventListener("click", () => {
+  }
+
+  handleSubmit(makeRequest, e).finally(() => {
+    closeModal(cardModal);
+    disableButton(cardSubmitBtn, settings);
```
You need to close modals only in `then` blocks after a **successful response** from the server so that when there is an error, it would be opened (the inputs will not be cleared and the user could send the data again to the server), and if you change the button text, you could see the change as well. **Please, correct it everywhere in the code**

gennady-bars replied:
You need to **disable** the button only in a `then` block after a **successful response** to be able to click it again if there is a server error.

gennady-bars reviewed (commented):
Everything is almost fine, but:

* You need to close modals only in `then` blocks after a **successful response** from the server so that when there is an error, it would be opened (the inputs will not be cleared and the user could send the data again to the server). **Please, correct it everywhere in the code**
* You need to **disable** the button only in a `then` block after a **successful response** to be able to click it again if there is a server error.

sun4205 committed:
disablebtn correction

sun4205 commented:
I added a new parameter, disableBtn, to the handleSubmit function. I now call disableButton only after a successful response. For the handleDeleteSubmit function, I set this parameter to false to prevent the button from being disabled after the deletion.
Thank you. :)

gennady-bars reviewed (approved):
You’ve done a great job!

Your project has been accepted. 

Good luck!

sun4205 commented:
Ye~~! Finally!! Thank you So Much!!!! :)
